server/app/assets/javascripts/session.ts

```diff
@@ -0,0 +1,396 @@
+      return parts.pop()?.split(';').shift() || null
+    }
+
+    return null
+  }
+
+  /**
+   * Main timer management method that:
+   * 1. Checks for immediate timeout conditions and logs out if needed
+   * 2. Shows inactivity warning once if its time has passed and not shown before
+   * 3. Shows total length warning once if its time has passed and not shown before
+   * 4. Sets timer for next future event (warning or timeout)
+   *
+   * Warning dialogs are shown only once per session and only one dialog
+   * can be visible at a time. If a warning was previously shown, it won't
+   * be shown again even if its time passes again.
+   */
+  private static checkAndSetTimer() {
+    const data = this.getTimeoutData()
+    if (!data) return
+
+    // Clear existing timer
+    if (this.timer) {
+      window.clearTimeout(this.timer)
+      this.timer = null
+    }
+
+    const now = Math.floor(Date.now() / 1000)
+
+    // 1. If there is an inactivityTimeout or totalTimeout that has passed, just logout
+    if (data.inactivityTimeout <= now || data.totalTimeout <= now) {
+      this.handleTimeout()
+      return
+    }
+
+    // If a warning is already being shown, don't show another one
+    if (this.inactivityWarningShown || this.totalLengthWarningShown) {
+      return
+    }
+
+    // 2 & 3. Show warnings if they haven't been shown before and their time has passed
+    if (!this.hasInactivityWarningBeenShown && data.inactivityWarning <= now) {
+      this.showWarning(WarningType.INACTIVITY)
+      this.hasInactivityWarningBeenShown = true
+      return
+    }
+
+    if (!this.hasTotalLengthWarningBeenShown && data.totalWarning <= now) {
+      this.showWarning(WarningType.TOTAL_LENGTH)
+      this.hasTotalLengthWarningBeenShown = true
+      return
+    }
+
+    // Set timers for future events
+    const timeouts = []
+
+    // Only add inactivity warning if it hasn't been shown yet
+    if (!this.hasInactivityWarningBeenShown && data.inactivityWarning > now) {
+      timeouts.push({
+        time: data.inactivityWarning,
+        action: () => {
+          this.showWarning(WarningType.INACTIVITY)
+          this.hasInactivityWarningBeenShown = true
+        },
+      })
+    }
+
+    // Only add total length warning if it hasn't been shown yet
+    if (!this.hasTotalLengthWarningBeenShown && data.totalWarning > now) {
+      timeouts.push({
+        time: data.totalWarning,
+        action: () => {
+          this.showWarning(WarningType.TOTAL_LENGTH)
+          this.hasTotalLengthWarningBeenShown = true
+        },
+      })
+    }
+
+    // Always add timeout events
+    timeouts.push({
+      time: data.inactivityTimeout,
+      action: () => this.handleTimeout(),
+    })
+
+    timeouts.push({
+      time: data.totalTimeout,
+      action: () => this.handleTimeout(),
+    })
+
+    // Sort by earliest time
+    timeouts.sort((a, b) => a.time - b.time)
+
+    if (timeouts.length > 0) {
```

Comment:
question (non-blocking): Is there ever a situation where we would get here and there would be no timeouts?

Reply:
You are right. Now since we are always adding the timers for timeout, this will always have legth 2 or more(2, 3, or 4).